tests/transforms/test_adjoint.py

```diff
@@ -85,3 +85,179 @@ def my_circuit():
     np.testing.assert_allclose(
         my_circuit(), np.array([-0.995707, 0.068644 + 6.209710e-02j]), atol=1e-6, rtol=1e-6
     )
+
+
+test_functions = [
+    lambda fn, *args, **kwargs: adjoint(fn)(*args, **kwargs),
+    lambda fn, *args, **kwargs: qml.inv(fn(*args, **kwargs)),
+]
+
+
+@pytest.mark.parametrize("fn", test_functions)
+class TestTemplateIntegration:
+    """Test that templates work correctly with the adjoint transform"""
+
+    def test_angle_embedding(self, fn):
+        """Test that the adjoint correctly inverts angle embedding"""
+        dev = qml.device("default.qubit", wires=3)
+        template = qml.templates.AngleEmbedding
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            template(features=weights, wires=[0, 1, 2])
+            fn(template, features=weights, wires=[0, 1, 2])
+            return qml.state()
+
+        weights = np.array([0.2, 0.5, 0.8])
+        res = circuit(weights)
+        assert len(np.nonzero(res)) == 1
+
+    def test_amplitude_embedding(self, fn):
+        """Test that the adjoint correctly inverts amplitude embedding"""
+        dev = qml.device("default.qubit", wires=3)
+        template = qml.templates.AmplitudeEmbedding
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            template(features=weights, wires=[0, 1, 2])
+            fn(template, features=weights, wires=[0, 1, 2])
+            return qml.state()
+
+        weights = np.array([0.2, 0.5, 0.8, 0.6, 0.1, 0.6, 0.1, 0.5]) / np.sqrt(1.92)
+        res = circuit(weights)
+        assert len(np.nonzero(res)) == 1
+
+    def test_basis_embedding(self, fn):
+        """Test that the adjoint correctly inverts basis embedding"""
+        dev = qml.device("default.qubit", wires=3)
+        template = qml.templates.BasisEmbedding
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            template(features=weights, wires=[0, 1, 2])
+            fn(template, features=weights, wires=[0, 1, 2])
+            return qml.state()
+
+        weights = np.array([1, 0, 1])
+        res = circuit(weights)
+        expected = np.zeros([2 ** 3])
+        expected[0] = 1.0
+        assert np.allclose(res, expected)

```

Comment:
Any reason to change the assert statement from the previous tests?

Reply:
I think I decided it was safer to check for the ground state explicitly, since any other basis state would also satisfy `assert len(np.nonzero(res)) == 1`